tools/analysis/passes/allfxevents/analysis.go

```diff
@@ -0,0 +1,278 @@
+// case for no-op or fake fxevent.Loggers, it ignores implementations that
+// handle none of the event types.
+//
+// This is meant for use within Fx only.
+package allfxevents
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"go/types"
+	"sort"
+	"strings"
+
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/inspect"
+	"golang.org/x/tools/go/ast/inspector"
+	"golang.org/x/tools/go/types/typeutil"
+)
+
+// Analyzer is a go/analysis compatible analyzer that verifies that all
+// fxevent.Loggers shipped with Fx handle all known Fx event types.
+var Analyzer = &analysis.Analyzer{
+	Name: "allfxevents",
+	Doc:  "check for unhandled fxevent.Events",
+	Run:  run,
+	Requires: []*analysis.Analyzer{
+		inspect.Analyzer,
+	},
+}
+
+var _filter = []ast.Node{
+	&ast.File{},
+	&ast.FuncDecl{},
+	&ast.CaseClause{},
+	&ast.TypeAssertExpr{},
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	fxeventPkg, ok := findPackage(pass.Pkg, "go.uber.org/fx/fxevent")
+	if !ok {
+		// If the package doesn't import fxevent, and itself isn't
+		// fxevent, then we don't need to run this pass.
+		return nil, nil
+	}
+
+	v := visitor{
+		Fxevent: inspectFxevent(fxeventPkg),
+		Fset:    pass.Fset,
+		Info:    pass.TypesInfo,
+		Report:  pass.Report,
+	}
+
+	pass.ResultOf[inspect.Analyzer].(*inspector.Inspector).Nodes(_filter, v.Visit)
+	return nil, nil
+}
+
+type visitor struct {
+	Fset    *token.FileSet
+	Info    *types.Info
+	Fxevent fxevent
+	Report  func(analysis.Diagnostic)
+
+	// types not yet referenced by this function
+	loggerType types.Type
+	funcEvents *typeSet
+}
+
+func (v *visitor) Visit(n ast.Node, push bool) (recurse bool) {
+	switch n := n.(type) {
+	case *ast.File:
+		if !push {
+			return false
+		}
+
+		// Don't run the linter on test files.
+		fname := v.Fset.File(n.Pos()).Name()
+		return !strings.HasSuffix(fname, "_test.go")
+
+	case *ast.FuncDecl:
+		if !push {
+			return v.funcDeclExit(n)
+		}
+		return v.funcDeclEnter(n)
+
+	case *ast.CaseClause:
+		if !push {
+			return false
+		}
+		for _, expr := range n.List {
+			t := v.Info.Types[expr].Type
+			if t != nil {
+				v.funcEvents.Remove(t)
+			}
+		}
+
+	case *ast.TypeAssertExpr:
+		if !push {
+			return false
+		}
+		t := v.Info.Types[n.Type].Type
+		if t != nil {
+			v.funcEvents.Remove(t)
+		}
+	}
+
+	return false
+}
+
+func (v *visitor) funcDeclEnter(n *ast.FuncDecl) bool {
+	// Skip top-level functions, and methods not named
+	// LogEvent.
+	if n.Recv == nil || n.Name.Name != "LogEvent" {
+		return false
+	}
+
+	// Skip types that don't implement fxevent.Logger.
+	t := v.Info.Types[n.Recv.List[0].Type].Type
+	if t == nil || !types.Implements(t, v.Fxevent.LoggerInterface) {
+		return false
+	}
+
+	// Each function declaration gets its own copy of the typeSet to track
+	// events in.
+	v.loggerType = t
+	v.funcEvents = v.Fxevent.Events.Clone()
+	return true
+}
+
+func (v *visitor) funcDeclExit(n *ast.FuncDecl) bool {
```

Comment:
`v.funcEvents.Len()` is repeatedly used. Can we pull it out to a variable and reuse it?
```suggestion
func (v *visitor) funcDeclExit(n *ast.FuncDecl) bool {
    nEvents := v.funcEvents.Len()
```